daemon/images/image_exporter.go

```diff
@@ -26,11 +26,9 @@ func (i *ImageService) PerformWithBaseFS(ctx context.Context, c *container.Conta
 		return err
 	}
 	defer func() {
-		if err != nil {
```

Comment:
Also trying to look what error this is checking for;

- the error produced at line 24 is unrelated (return early)
- the error produced  line 35 (`rwlayer.Mount`) is the one handled here (i.e., it the mount failed, we will try to release the layer)
- but errors returned by `fn(basefs)` are not considered.

So, if the intent was to release when the mount failed, this wouldn't have to be a defer, and could be put in the error-handling block here; https://github.com/moby/moby/blob/5098132d84d3c20afd05453e430b5029e770fcc1/daemon/images/image_exporter.go#L37-L40

But indeed, if releasing only happens on an error, then there should likely be other code handling the release (which may be outside of this function, or as part of the `fn(basefs)`.

If it's the caller's responsibility (or the `fn`'s responsibility) to release the layer, that should at least be documented 😅

Reply:
`PerformWithBaseFS` looks to be used in 1 location, and AFAICS that doesn't release the layer; https://github.com/moby/moby/blob/5098132d84d3c20afd05453e430b5029e770fcc1/daemon/export.go#L45-L58


It's also called as part of the containerd image service, but that's a different implementation; https://github.com/moby/moby/blob/5098132d84d3c20afd05453e430b5029e770fcc1/daemon/containerd/image_changes.go#L31-L36

Reply:
> But indeed, if releasing only happens on an error, then there should likely be other code handling the release (which may be outside of this function, or as part of the fn(basefs).

> If it's the caller's responsibility (or the fn's responsibility) to release the layer, that should at least be documented 😅


IMHO if calling `PerformWithBaseFS` is "locking/leaking" the container layer, I guess it's less error prone to release it when the function completes :shrug: